pkg/operator/clustermembercontroller/clustermembercontroller.go

```diff
@@ -93,83 +118,220 @@ func (c *ClusterMemberController) reconcileMembers(ctx context.Context, recorder
+	// Set desired cluster size
+	// TODO(haseeb): Does this replica count config ever change, and if so, should we react to that?
+	if c.mastersReplicaCount == 0 {
+		mastersReplicaCount, err := ceohelpers.GetMastersReplicaCount(c.configMapLister)
+		if err != nil {
+			return fmt.Errorf("failed to get control-plane replica count: %w", err)
+		}
+		c.mastersReplicaCount = mastersReplicaCount
 	}
 
-	etcdHost, err := c.getEtcdPeerHostToScale(podToAdd)
+	// Add a learner member if next peer found
+	peerURL, err := c.getEtcdPeerURLToAdd(ctx)
 	if err != nil {
-		return fmt.Errorf("could not get etcd peer host :%w", err)
+		return fmt.Errorf("could not get etcd peerURL to add :%w", err)
 	}
-	err = c.etcdClient.MemberAdd(ctx, fmt.Sprintf("https://%s:2380", etcdHost))
-	if err != nil {
-		return fmt.Errorf("could not add member :%w", err)
+	if peerURL != "" {
+		if err = c.etcdClient.MemberAddAsLearner(ctx, peerURL); err != nil {
+			return fmt.Errorf("failed to add learner member: %w", err)
+		}
+	}
+
+	// Promote a promotable learner member if there is room for more voting members
+	if err := c.ensureEtcdLearnerPromotion(ctx, recorder); err != nil {
+		return fmt.Errorf("failed to ensure learner promotion: %w", err)
 	}
+
 	return nil
 }
 
-func (c *ClusterMemberController) getEtcdPodToAddToMembership(ctx context.Context) (*corev1.Pod, error) {
-	// list etcd member pods
-	pods, err := c.podLister.List(labels.Set{"app": "etcd"}.AsSelector())
+// getEtcdPeerURLToAdd returns a PeerURL of the next etcd instance to be added as a learner to the cluster. If
+// no elegible member is found return an empty string.
+func (c *ClusterMemberController) getEtcdPeerURLToAdd(ctx context.Context) (string, error) {
+	members, err := c.etcdClient.MemberList(ctx)
 	if err != nil {
-		return nil, err
+		return "", err
 	}
 
-	// go through the list of all pods, pick one peerFQDN to return from unready pods
-	// and collect dns resolution errors on the way.
-	for _, pod := range pods {
-		if !strings.HasPrefix(pod.Name, "etcd-") {
+	// For a desired size of N voting members in a cluster (specified in the install-config)
+	// we can only add a new learner member if we're under 2N total members (voting+learners).
+	// This abides by the current max limit of N learner members.
+	if len(members) >= 2*c.mastersReplicaCount {
+		klog.V(4).Infof("Skipping learner member addition since the current membership size (%v) exceeds the max learners limit of %v ", len(members), 2*c.mastersReplicaCount)
+		return "", nil
+	}
+
+	nodes, err := c.masterNodeLister.List(c.masterNodeSelector)
+	if err != nil {
+		return "", err
+	}
+
+	isFunctional, err := c.machineAPIChecker.IsFunctional()
+	if err != nil {
+		return "", fmt.Errorf("failed to determine Machine API availability: %v", err)
+	}
+
+	for _, node := range nodes {
+		if node.DeletionTimestamp != nil {
+			klog.V(4).Infof("Ignoring node (%s) for new member addition as it's pending deletion", node.Name)
```

Comment:
I think our default is `V(2)`. This info seems important, would it make sense to decrease the log level.